topdown/tokens_verify.go

```diff
@@ -0,0 +1,127 @@
+// Copyright 2017 The OPA Authors.  All rights reserved.
+// Use of this source code is governed by an Apache2
+// license that can be found in the LICENSE file.
+
+package topdown
+
+import (
+	"crypto"
+	"crypto/rsa"
+	"crypto/sha256"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/open-policy-agent/opa/ast"
+	"github.com/open-policy-agent/opa/topdown/builtins"
+)
+
+// Implements RS256 JWT signature verification
+func builtinJWTVerifyRS256(a ast.Value, b ast.Value) (ast.Value, error) {
+
+	// Process the token
+	astToken, err := builtins.StringOperand(a, 1)
+	if err != nil {
+		return nil, err
+	}
+	token := string(astToken)
+	if !strings.Contains(token, ".") {
+		return nil, errors.New("encoded JWT has no period separators")
+	}
+
+	parts := strings.Split(token, ".")
+	if len(parts) != 3 {
+		return nil, fmt.Errorf("encoded JWT must have 3 sections, found %d", len(parts))
+	}
+
+	result := make([][]byte, len(parts))
+	for i, part := range parts {
+		if result[i], err = decode(part); err != nil {
+			return nil, err
+		}
+	}
+
+	header, payload, signature := result[0], result[1], result[2]
+	headerPayload := []byte(serialize(header, payload))
+
+	// Process PEM encoded certificate input
+	astCertificate, err := builtins.StringOperand(b, 2)
+	if err != nil {
+		return nil, err
+	}
+	certificate := string(astCertificate)
+
+	block, rest := pem.Decode([]byte(certificate))
+
+	if block == nil || block.Type != "CERTIFICATE" || len(rest) > 0 {
+		return nil, fmt.Errorf("failed to decode PEM block containing certificate")
+	}
+
+	cert, err := x509.ParseCertificate(block.Bytes)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse PEM certificate %v", err)
```

Comment:
See `errors.Wrapf` for returning wrapped errors. This way the embedded error formatting is more consistent. For example, I would use `errors.Wrapf(err, "PEM parse error")`.

Reply:
Used errors.Wrap